plugins/Vectorscope/VectorView.cpp

```diff
@@ -0,0 +1,284 @@
+	const int displayBottom = height() - 2;
+	const int displayLeft = 2;
+	const int displayRight = width() - 2;
+	const int displayWidth = displayRight - displayLeft;
+	const int displayHeight = displayBottom - displayTop;
+
+	const float centerX = displayLeft + (displayWidth / 2.f);
+	const float centerY = displayTop + (displayWidth / 2.f);
+
+	const int margin = 4;
+	const int gridCorner = 30;
+
+	// Setup QPainter and font sizes
+	QPainter painter(this);
+	painter.setRenderHint(QPainter::Antialiasing, true);
+
+	QFont normalFont, boldFont;
+	boldFont.setPixelSize(26);
+	boldFont.setBold(true);
+	const int labelWidth = 26;
+	const int labelHeight = 26;
+
+	bool hq = m_controls->m_highQualityModel.value();
+
+	// Clear display buffer if quality setting was changed
+	if (hq != m_oldHQ)
+	{
+		m_oldHQ = hq;
+		for (std::size_t i = 0; i < m_displayBuffer.size(); i++)
+		{
+			m_displayBuffer.data()[i] = 0;
+		}
+	}
+
+	// Dim stored image based on persistence setting and elapsed time.
+	// Update period is limited to 50 ms (20 FPS) for non-HQ mode and 10 ms (100 FPS) for HQ mode.
+	const unsigned int currentTimestamp = std::chrono::duration_cast<std::chrono::milliseconds>
+	(
+		std::chrono::high_resolution_clock::now().time_since_epoch()
+	).count();
+	const unsigned int elapsed = currentTimestamp - m_persistTimestamp;
+	const unsigned int threshold = hq ? 10 : 50;
+	if (elapsed > threshold)
+	{
+		m_persistTimestamp = currentTimestamp;
+		// Non-HQ mode uses half the resolution → use limited buffer space.
+		const std::size_t useableBuffer = hq ? m_displayBuffer.size() : m_displayBuffer.size() / 4;
+		// The knob value is interpreted on log. scale, otherwise the effect would ramp up too slowly.
+		// Persistence value specifies fraction of light intensity that remains after 10 ms.
+		// → Compensate it based on elapsed time (exponential decay).
+		const float persist = log10(1 + 9 * m_controls->m_persistenceModel.value());
+		const float persistPerFrame = pow(persist, elapsed / 10.f);
+		// Note that for simplicity and performance reasons, this implementation only dims all stored
+		// values by a given factor. A true simulation would also do the inverse of desaturation that
+		// occurs in high-intensity traces in HQ mode.
+		for (std::size_t i = 0; i < useableBuffer; i++)
+		{
+			m_displayBuffer.data()[i] = m_displayBuffer.data()[i] * persistPerFrame;
+		}
+	}
+
+	// Get new samples from the lockless input FIFO buffer
+	auto inBuffer = m_bufferReader.read_max(m_inputBuffer->capacity());
+	std::size_t frameCount = inBuffer.size();
+
+	// Draw new points on top
+	float left, right;
+	int x, y;
+
+	const bool logScale = m_controls->m_logarithmicModel.value();
+	const unsigned short activeSize = hq ? m_displaySize : m_displaySize / 2;
+
+	// Helper lambda functions for better readability
+	// Make sure pixel stays within display bounds:
+	auto saturate = [=](short pixelPos) {return fmax(fmin(pixelPos, activeSize - 1), 0);};
```

Comment:
FYI, you can use `qBound()` for this purpose. Also, remind that `fmax` and `fmin` return `double` in this case. Maybe you'd better using `std::max` and `std::min` instead.

Reply:
You may take care of the type of integer literals as well.

Reply:
> FYI, you can use qBound() for this purpose

Ah, nice. I was searching for something like that in the `std` library and `cmath` and did not find anything. 

What do you mean about the integers? I notice that some may be too big; did you mean that?

Reply:
I mean something like `(short)0` vs. `0`.